relay-base-schema/src/metrics/mri.rs

```diff
@@ -0,0 +1,471 @@
+use std::fmt;
+use std::{borrow::Cow, error::Error};
+
+use crate::metrics::MetricUnit;
+use serde::{Deserialize, Serialize};
+
+/// The type of a [`MetricResourceIdentifier`], determining its aggregation and evaluation.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub enum MetricType {
+    /// Counts instances of an event.
+    ///
+    /// Counters can be incremented and decremented. The default operation is to increment a counter
+    /// by `1`, although increments by larger values are equally possible.
+    ///
+    /// Counters are declared as `"c"`. Alternatively, `"m"` is allowed.
+    Counter,
+    /// Builds a statistical distribution over values reported.
+    ///
+    /// Based on individual reported values, distributions allow to query the maximum, minimum, or
+    /// average of the reported values, as well as statistical quantiles. With an increasing number
+    /// of values in the distribution, its accuracy becomes approximate.
+    ///
+    /// Distributions are declared as `"d"`. Alternatively, `"d"` and `"ms"` are allowed.
+    Distribution,
+    /// Counts the number of unique reported values.
+    ///
+    /// Sets allow sending arbitrary discrete values, including strings, and store the deduplicated
+    /// count. With an increasing number of unique values in the set, its accuracy becomes
+    /// approximate. It is not possible to query individual values from a set.
+    ///
+    /// Sets are declared as `"s"`.
+    Set,
+    /// Stores absolute snapshots of values.
+    ///
+    /// In addition to plain [counters](Self::Counter), gauges store a snapshot of the maximum,
+    /// minimum and sum of all values, as well as the last reported value.
+    ///
+    /// Gauges are declared as `"g"`.
+    Gauge,
+}
+
+impl MetricType {
+    /// Return the shortcode for this metric type.
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            MetricType::Counter => "c",
+            MetricType::Distribution => "d",
+            MetricType::Set => "s",
+            MetricType::Gauge => "g",
+        }
+    }
+}
+
+impl fmt::Display for MetricType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(self.as_str())
+    }
+}
+
+impl std::str::FromStr for MetricType {
+    type Err = ParseMetricError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(match s {
+            "c" | "m" => Self::Counter,
+            "h" | "d" | "ms" => Self::Distribution,
+            "s" => Self::Set,
+            "g" => Self::Gauge,
+            _ => return Err(ParseMetricError),
+        })
+    }
+}
+
+relay_common::impl_str_serde!(MetricType, "a metric type string");
+
+/// An error returned when metrics or MRIs cannot be parsed.
+#[derive(Clone, Copy, Debug)]
+pub struct ParseMetricError;
```

Comment:
Was `pub struct ParseMetricError(pub(crate) ());` but this is kinda awkward since we need the error as well in `Bucket` (`relay-metrics`)